File: python/mirage/workspace/executor/builtins/condition.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import re
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Union

from mirage.io import IOResult
from mirage.io.types import ByteSource, materialize
from mirage.shell.arith import ArithError, evaluate_arith
from mirage.shell.errors import ExitSignal
from mirage.types import FileStat, FileType, PathSpec
from mirage.utils.fnmatch import fnmatch
from mirage.utils.path import resolve_path, resolve_symlinks
from mirage.workspace.executor.builtins.scope import _scope_path, _to_scope
from mirage.workspace.mount.namespace import Namespace
from mirage.workspace.session import Session
from mirage.workspace.types import ExecutionNode

CondNode = Union["CondWord", "CondUnary", "CondBinary", "CondNot", "CondAnd",
                 "CondOr"]


@dataclass(frozen=True, slots=True)
class CondWord:
    value: str


@dataclass(frozen=True, slots=True)
class CondUnary:
    op: str
    operand: str


@dataclass(frozen=True, slots=True)
class CondBinary:
    left: str
    op: str
    right: str
    # True when the right side was quoted: `[[ x == "a*" ]]` compares
    # literally while the unquoted form pattern-matches.
    right_literal: bool = False


@dataclass(frozen=True, slots=True)
class CondNot:
    inner: CondNode


@dataclass(frozen=True, slots=True)
class CondAnd:
    left: CondNode
    right: CondNode


@dataclass(frozen=True, slots=True)
class CondOr:
    left: CondNode
    right: CondNode


class CondError(Exception):
    """A test/[/[[ usage error: bash prints the message and returns 2.

    Args:
        message (str): diagnostic without trailing newline.
    """

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message = message


@dataclass(frozen=True, slots=True)
class CondContext:
    dispatch: Callable[..., Any]
    namespace: Namespace
    session: Session
    name: str


_STRING_BINARY = frozenset({"=", "==", "!="})
_NUMERIC_BINARY = frozenset({"-eq", "-ne", "-lt", "-le", "-gt", "-ge"})
_FILE_PAIR_BINARY = frozenset({"-nt", "-ot", "-ef"})
_STRING_UNARY = frozenset({"-n", "-z"})
_FILE_UNARY = frozenset({"-e", "-f", "-d", "-s", "-r", "-w", "-x", "-L", "-h"})
# Real GNU operators mirage cannot answer (no pipe/socket/tty/owner model);
# failing loudly beats the silent-false this module used to produce.
_UNSUPPORTED_UNARY = frozenset(
    {"-p", "-S", "-b", "-c", "-g", "-k", "-u", "-O", "-G", "-N", "-t"})
_BINARY_OPS = _STRING_BINARY | _NUMERIC_BINARY | _FILE_PAIR_BINARY
_UNARY_OPS = _STRING_UNARY | _FILE_UNARY | _UNSUPPORTED_UNARY


def _operand_scope(ctx: CondContext, val: str | PathSpec) -> PathSpec:
    """Resolve a file operand to an addressable scope.

    Args:
        ctx (CondContext): evaluation context.
        val (str | PathSpec): operand as typed or classified.
    """
    if isinstance(val, PathSpec):
        return val
    resolved = resolve_path(val, ctx.session.cwd)
    resolved = resolve_symlinks(resolved, ctx.namespace.symlink_targets())
    return _to_scope(resolved)


async def _path_kind(
        ctx: CondContext,
        val: str | PathSpec) -> tuple[str | None, FileStat | None]:
    """Resolve an operand to 'dir' / 'file' / None plus its stat.

    Symlinks are followed first (test -e/-f/-d act on the target); a
    stat that names a directory type answers directly, otherwise a
    readdir probe catches backends whose stat cannot see directories.
    The probe demands a non-empty listing: prefix stores (s3, gridfs,
    hf, nextcloud) list a missing path as [] instead of raising, and
    they cannot hold an empty directory anyway.

    Args:
        ctx (CondContext): evaluation context.
        val (str | PathSpec): operand as typed or classified.
    """
    scope = _operand_scope(ctx, val)
    try:
        stat, _ = await ctx.dispatch("stat", scope)
    except (FileNotFoundError, ValueError, NotADirectoryError):
        stat = None
    if stat is not None:
        if stat.type == FileType.DIRECTORY:
            return "dir", stat
        return "file", stat
    try:
        entries, _ = await ctx.dispatch("readdir", scope)
    except (FileNotFoundError, ValueError, NotADirectoryError):
        return None, None
    if entries:
        return "dir", None
    return None, None


async def _apply_unary(ctx: CondContext, op: str, val: str | PathSpec) -> bool:
    """Evaluate a unary operator.

    Args:
        ctx (CondContext): evaluation context.
        op (str): operator token, e.g. ``-n`` or ``-e``.
        val (str | PathSpec): operand.
    """
    text = _scope_path(val)
    if op == "-n":
        return text != ""
    if op == "-z":
        return text == ""
    if op in ("-L", "-h"):
        resolved = resolve_path(text, ctx.session.cwd)
        return ctx.namespace.is_link(resolved)
    if op in _FILE_UNARY:
        if not isinstance(val, PathSpec) and not text:
            return False
        kind, stat = await _path_kind(ctx, val)
        if op == "-e":
            return kind is not None
        if op == "-f":
            return kind == "file"
        if op == "-d":
            return kind == "dir"
        if op == "-s":
            if kind == "dir":
                return True
            if kind != "file" or stat is None:
                return False
            if stat.size is not None:
                return stat.size > 0
            # API backends (dropbox, gdrive, box) stat freshly written
            # empty files as size-unknown; only a read can answer, and
            # the prefetch TTL cache keeps repeat tests cheap.
            data, _ = await ctx.dispatch("read", _operand_scope(ctx, val))
            return len(await materialize(data)) > 0
        if op in ("-r", "-w"):
            # Mirage has no per-user access model: whatever exists in a
            # mount is readable and writable through it.
            return kind is not None
        if op == "-x":
            if kind == "dir":
                return True
            if kind != "file" or stat is None:
                return False
            return stat.mode is not None and bool(stat.mode & 0o111)
    if op in _UNSUPPORTED_UNARY:
        raise CondError(f"{ctx.name}: {op}: unsupported operator")
    raise CondError(f"{ctx.name}: {op}: unary operator expected")


def _to_int(ctx: CondContext, text: str) -> int:
    """Parse a test integer operand, with bash's diagnostic.

    Args:
        ctx (CondContext): evaluation context.
        text (str): operand text.
    """
    try:
        return int(text.strip())
    except ValueError:
        raise CondError(f"{ctx.name}: {text}: integer expression expected")


async def _apply_binary(ctx: CondContext, left: str | PathSpec, op: str,
                        right: str | PathSpec) -> bool:
    """Evaluate a test/[ binary operator (literal string semantics).

    Args:
        ctx (CondContext): evaluation context.
        left (str | PathSpec): left operand.
        op (str): operator token.
        right (str | PathSpec): right operand.
    """
    lt = _scope_path(left)
    rt = _scope_path(right)
    if op in ("=", "=="):
        return lt == rt
    if op == "!=":
        return lt != rt
    if op in _NUMERIC_BINARY:
        li = _to_int(ctx, lt)
        ri = _to_int(ctx, rt)
        return _compare_ints(op, li, ri)
    if op in _FILE_PAIR_BINARY:
        raise CondError(f"{ctx.name}: {op}: unsupported operator")
    raise CondError(f"{ctx.name}: {op}: binary operator expected")


def _compare_ints(op: str, li: int, ri: int) -> bool:
    """Apply a numeric comparison operator.

    Args:
        op (str): one of -eq -ne -lt -le -gt -ge.
        li (int): left value.
        ri (int): right value.
    """
    if op == "-eq":
        return li == ri
    if op == "-ne":
        return li != ri
    if op == "-lt":
        return li < ri
    if op == "-le":
        return li <= ri
    if op == "-gt":
        return li > ri
    return li >= ri


async def _eval_one(ctx: CondContext, arg: str | PathSpec) -> bool:
    """One-argument rule: true when the operand is non-empty.

    Args:
        ctx (CondContext): evaluation context.
        arg (str | PathSpec): sole operand.
    """
    return _scope_path(arg) != ""


async def _eval_two(ctx: CondContext, argv: list[str | PathSpec]) -> bool:
    """Two-argument rule: ``!`` or a unary operator.

    Args:
        ctx (CondContext): evaluation context.
        argv (list[str | PathSpec]): exactly two operands.
    """
    first = _scope_path(argv[0])
    if first == "!":
        return not await _eval_one(ctx, argv[1])
    if first in _UNARY_OPS:
        return await _apply_unary(ctx, first, argv[1])
    raise CondError(f"{ctx.name}: {first}: unary operator expected")


async def _eval_three(ctx: CondContext, argv: list[str | PathSpec]) -> bool:
    """Three-argument rule: binary op, ``-a``/``-o``, ``!``, or parens.

    Args:
        ctx (CondContext): evaluation context.
        argv (list[str | PathSpec]): exactly three operands.
    """
    mid = _scope_path(argv[1])
    if mid == "-a":
        return await _eval_one(ctx, argv[0]) and await _eval_one(ctx, argv[2])
    if mid == "-o":
        return await _eval_one(ctx, argv[0]) or await _eval_one(ctx, argv[2])
    if mid in _BINARY_OPS:
        return await _apply_binary(ctx, argv[0], mid, argv[2])
    first = _scope_path(argv[0])
    if first == "!":
        return not await _eval_two(ctx, argv[1:])
    if first == "(" and _scope_path(argv[2]) == ")":
        return await _eval_one(ctx, argv[1])
    raise CondError(f"{ctx.name}: {mid}: binary operator expected")


async def _eval_four(ctx: CondContext, argv: list[str | PathSpec]) -> bool:
    """Four-argument rule: leading ``!`` or parens, else expr parser.

    Args:
        ctx (CondContext): evaluation context.
        argv (list[str | PathSpec]): exactly four operands.
    """
    first = _scope_path(argv[0])
    if first == "!":
        return not await _eval_three(ctx, argv[1:])
    if first == "(" and _scope_path(argv[3]) == ")":
        return await _eval_two(ctx, argv[1:3])
    parser = _ExprParser(ctx, argv)
    return await parser.run()


class _ExprParser:
    """Recursive-descent ``test`` expression parser (>4 args, GNU expr
    grammar: or -> and -> term with ``!`` and parentheses)."""

    def __init__(self, ctx: CondContext, argv: list[str | PathSpec]) -> None:
        self.ctx = ctx
        self.argv = argv
        self.pos = 0

    def _peek(self, offset: int = 0) -> str | None:
        i = self.pos + offset
        if i < len(self.argv):
            return _scope_path(self.argv[i])
        return None

    async def run(self) -> bool:
        result = await self._or_expr()
        if self.pos != len(self.argv):
            raise CondError(f"{self.ctx.name}: too many arguments")
        return result

    async def _or_expr(self) -> bool:
        result = await self._and_expr()
        while self._peek() == "-o":
            self.pos += 1
            right = await self._and_expr()
            result = result or right
        return result

    async def _and_expr(self) -> bool:
        result = await self._term()
        while self._peek() == "-a":
            self.pos += 1
            right = await self._term()
            result = result and right
        return result

    async def _term(self) -> bool:
        tok = self._peek()
        if tok is None:
            raise CondError(f"{self.ctx.name}: argument expected")
        if tok == "!":
            self.pos += 1
            return not await self._term()
        if tok == "(":
            self.pos += 1
            result = await self._or_expr()
            if self._peek() != ")":
                raise CondError(f"{self.ctx.name}: `)' expected")
            self.pos += 1
            return result
        nxt = self._peek(1)
        if nxt in _BINARY_OPS and self._peek(2) is not None:
            left = self.argv[self.pos]
            right = self.argv[self.pos + 2]
            self.pos += 3
            return await _apply_binary(self.ctx, left, nxt, right)
        if tok in _UNARY_OPS and nxt is not None and nxt not in ("-a", "-o"):
            operand = self.argv[self.pos + 1]
            self.pos += 2
            return await _apply_unary(self.ctx, tok, operand)
        self.pos += 1
        return tok != ""


async def eval_flat(ctx: CondContext, argv: list[str | PathSpec]) -> bool:
    """Evaluate a flat ``test``/``[`` argument list with bash's arity
    rules.

    Args:
        ctx (CondContext): evaluation context.
        argv (list[str | PathSpec]): operands, brackets excluded.
    """
    n = len(argv)
    if n == 0:
        return False
    if n == 1:
        return await _eval_one(ctx, argv[0])
    if n == 2:
        return await _eval_two(ctx, argv)
    if n == 3:
        return await _eval_three(ctx, argv)
    if n == 4:
        return await _eval_four(ctx, argv)
    parser = _ExprParser(ctx, argv)
    return await parser.run()


async def eval_cond(ctx: CondContext, node: CondNode) -> bool:
    """Evaluate a structured ``[[ ]]`` expression tree.

    Args:
        ctx (CondContext): evaluation context.
        node (CondNode): parsed condition.
    """
    if isinstance(node, CondAnd):
        return (await eval_cond(ctx, node.left)
                and await eval_cond(ctx, node.right))
    if isinstance(node, CondOr):
        return (await eval_cond(ctx, node.left)
                or await eval_cond(ctx, node.right))
    if isinstance(node, CondNot):
        return not await eval_cond(ctx, node.inner)
    if isinstance(node, CondUnary):
        if node.op not in _UNARY_OPS:
            raise CondError("mirage: conditional unary operator expected")
        return await _apply_unary(ctx, node.op, node.operand)
    if isinstance(node, CondBinary):
        return await _eval_cond_binary(ctx, node)
    return node.value != ""


async def _eval_cond_binary(ctx: CondContext, node: CondBinary) -> bool:
    """Evaluate a ``[[ ]]`` binary: pattern/regex/string/arith semantics.

    Args:
        ctx (CondContext): evaluation context.
        node (CondBinary): binary condition.
    """
    if node.op in ("=", "=="):
        if node.right_literal:
            return node.left == node.right
        return fnmatch(node.left, node.right)
    if node.op == "!=":
        if node.right_literal:
            return node.left != node.right
        return not fnmatch(node.left, node.right)
    if node.op == "=~":
        pattern = re.escape(node.right) if node.right_literal else node.right
        try:
            match = re.search(pattern, node.left)
        except re.error:
            raise CondError("mirage: syntax error in conditional expression")
        if match is None:
            return False
        groups = [g if g is not None else "" for g in match.groups()]
        ctx.session.arrays["BASH_REMATCH"] = [match.group(0), *groups]
        return True
    if node.op == "<":
        return node.left < node.right
    if node.op == ">":
        return node.left > node.right
    if node.op in _NUMERIC_BINARY:
        # [[ evaluates numeric operands as arithmetic: variables
        # resolve, expressions compute, bare unset words are 0.
        try:
            li, _ = evaluate_arith(node.left, ctx.session.env)
            ri, _ = evaluate_arith(node.right, ctx.session.env)
        except ArithError:
            raise CondError("mirage: syntax error in conditional expression")
        return _compare_ints(node.op, li, ri)
    if node.op in _FILE_PAIR_BINARY:
        raise CondError(f"{ctx.name}: {node.op}: unsupported operator")
    raise CondError("mirage: conditional binary operator expected")


async def handle_test(
    dispatch: Callable[..., Any],
    namespace: Namespace,
    args: list[str | PathSpec] | CondNode,
    session: Session,
    name: str = "test",
) -> tuple[ByteSource | None, IOResult, ExecutionNode]:
    """Evaluate test/[ (flat argv) or [[ (condition tree).

    Args:
        dispatch (Callable): op dispatcher for file probes.
        namespace (Namespace): addressing authority (symlink table).
        args (list[str | PathSpec] | CondNode): flat operands for
            test/[, a CondNode tree for [[.
        session (Session): session for cwd, env, and BASH_REMATCH.
        name (str): invocation name for diagnostics: "test", "[", "[[".
    """
    ctx = CondContext(dispatch=dispatch,
                      namespace=namespace,
                      session=session,
                      name=name)
    try:
        if isinstance(args, list):
            result = await eval_flat(ctx, args)
        else:
            result = await eval_cond(ctx, args)
    except CondError as err:
        stderr = (err.message + "\n").encode()
        if name == "[[":
            # A bad [[ ]] operator is a bash PARSE error: the whole
            # input line dies, not just this command.
            raise ExitSignal(2, stderr=stderr, contained_code=2)
        return None, IOResult(exit_code=2,
                              stderr=stderr), ExecutionNode(command="test",
                                                            exit_code=2,
                                                            stderr=stderr)
    code = 0 if result else 1
    return None, IOResult(exit_code=code), ExecutionNode(command="test",
                                                         exit_code=code)

File: typescript/packages/core/src/workspace/executor/builtins/condition.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { evaluateArith } from '../../../shell/arith.ts'
import { ArithError, ExitSignal } from '../../../shell/errors.ts'
import { fnmatch } from '../../../utils/fnmatch.ts'
import { resolvePath, resolveSymlinks } from '../../../utils/path.ts'
import { IOResult } from '../../../io/types.ts'
import type { FileStat } from '../../../types.ts'
import { FileType, PathSpec } from '../../../types.ts'
import type { Namespace } from '../../mount/namespace/namespace.ts'
import type { Session } from '../../session/session.ts'
import { ExecutionNode } from '../../types.ts'
import type { DispatchFn } from '../cross_mount.ts'
import { toScope, scopePath } from './scope.ts'
import type { Result } from './scope.ts'

export type CondNode =
  | { kind: 'word'; value: string }
  | { kind: 'unary'; op: string; operand: string }
  | { kind: 'binary'; left: string; op: string; right: string; rightLiteral: boolean }
  | { kind: 'not'; inner: CondNode }
  | { kind: 'and'; left: CondNode; right: CondNode }
  | { kind: 'or'; left: CondNode; right: CondNode }

/** A test/[/[[ usage error: bash prints the message and returns 2. */
class CondError extends Error {}

interface CondContext {
  dispatch: DispatchFn
  namespace: Namespace
  session: Session
  name: string
}

const STRING_BINARY = new Set(['=', '==', '!='])
const NUMERIC_BINARY = new Set(['-eq', '-ne', '-lt', '-le', '-gt', '-ge'])
const FILE_PAIR_BINARY = new Set(['-nt', '-ot', '-ef'])
const STRING_UNARY = new Set(['-n', '-z'])
const FILE_UNARY = new Set(['-e', '-f', '-d', '-s', '-r', '-w', '-x', '-L', '-h'])
// Real GNU operators mirage cannot answer (no pipe/socket/tty/owner model);
// failing loudly beats the silent-false this module used to produce.
const UNSUPPORTED_UNARY = new Set([
  '-p',
  '-S',
  '-b',
  '-c',
  '-g',
  '-k',
  '-u',
  '-O',
  '-G',
  '-N',
  '-t',
])
const BINARY_OPS = new Set([...STRING_BINARY, ...NUMERIC_BINARY, ...FILE_PAIR_BINARY])
const UNARY_OPS = new Set([...STRING_UNARY, ...FILE_UNARY, ...UNSUPPORTED_UNARY])

function isMissError(exc: unknown): boolean {
  const code = (exc as { code?: string }).code
  if (code === 'ENOENT' || code === 'ENOTDIR' || code === 'EISDIR') return true
  const msg = exc instanceof Error ? exc.message : String(exc)
  return /not found|no such file|not a directory|is a directory/i.test(msg)
}

/**
 * Resolve an operand to 'dir' / 'file' / null plus its stat. Symlinks are
 * followed first; a stat naming a directory type answers directly, and a
 * readdir probe catches backends whose stat cannot see directories.
 */
async function pathKind(
  ctx: CondContext,
  val: string | PathSpec,
): Promise<['dir' | 'file' | null, FileStat | null]> {
  let scope: PathSpec
  if (val instanceof PathSpec) {
    scope = val
  } else {
    let resolved = resolvePath(val, ctx.session.cwd)
    resolved = resolveSymlinks(resolved, ctx.namespace.symlinkTargets())
    scope = toScope(resolved)
  }
  let stat: FileStat | null = null
  try {
    const [s] = await ctx.dispatch('stat', scope)
    stat = s as FileStat | null
  } catch (exc) {
    if (!isMissError(exc)) throw exc
  }
  if (stat !== null) {
    if (stat.type === FileType.DIRECTORY) return ['dir', stat]
    return ['file', stat]
  }
  try {
    await ctx.dispatch('readdir', scope)
    return ['dir', null]
  } catch (exc) {
    if (!isMissError(exc)) throw exc
    return [null, null]
  }
}

async function applyUnary(ctx: CondContext, op: string, val: string | PathSpec): Promise<boolean> {
  const text = scopePath(val)
  if (op === '-n') return text !== ''
  if (op === '-z') return text === ''
  if (op === '-L' || op === '-h') {
    const resolved = resolvePath(text, ctx.session.cwd)
    return ctx.namespace.isLink(resolved)
  }
  if (FILE_UNARY.has(op)) {
    if (!(val instanceof PathSpec) && text === '') return false
    const [kind, stat] = await pathKind(ctx, val)
    if (op === '-e') return kind !== null
    if (op === '-f') return kind === 'file'
    if (op === '-d') return kind === 'dir'
    if (op === '-s') {
      if (kind === 'dir') return true
      if (kind !== 'file' || stat === null) return false
      // Unknown API-backed sizes count as non-empty: the file exists
      // and hydration would be a full fetch per test.
      return stat.size === null || stat.size > 0
    }
    if (op === '-r' || op === '-w') {
      // Mirage has no per-user access model: whatever exists in a
      // mount is readable and writable through it.
      return kind !== null
    }
    if (op === '-x') {
      if (kind === 'dir') return true
      if (kind !== 'file' || stat === null) return false
      return stat.mode !== null && (stat.mode & 0o111) !== 0
    }
  }
  if (UNSUPPORTED_UNARY.has(op)) {
    throw new CondError(`${ctx.name}: ${op}: unsupported operator`)
  }
  throw new CondError(`${ctx.name}: ${op}: unary operator expected`)
}

function toInt(ctx: CondContext, text: string): bigint {
  const trimmed = text.trim()
  if (!/^[+-]?\d+$/.test(trimmed)) {
    throw new CondError(`${ctx.name}: ${text}: integer expression expected`)
  }
  return BigInt(trimmed)
}

function compareInts(op: string, li: bigint, ri: bigint): boolean {
  if (op === '-eq') return li === ri
  if (op === '-ne') return li !== ri
  if (op === '-lt') return li < ri
  if (op === '-le') return li <= ri
  if (op === '-gt') return li > ri
  return li >= ri
}

function applyBinary(
  ctx: CondContext,
  left: string | PathSpec,
  op: string,
  right: string | PathSpec,
): boolean {
  const lt = scopePath(left)
  const rt = scopePath(right)
  if (op === '=' || op === '==') return lt === rt
  if (op === '!=') return lt !== rt
  if (NUMERIC_BINARY.has(op)) {
    const li = toInt(ctx, lt)
    const ri = toInt(ctx, rt)
    return compareInts(op, li, ri)
  }
  if (FILE_PAIR_BINARY.has(op)) {
    throw new CondError(`${ctx.name}: ${op}: unsupported operator`)
  }
  throw new CondError(`${ctx.name}: ${op}: binary operator expected`)
}

function evalOne(arg: string | PathSpec): boolean {
  return scopePath(arg) !== ''
}

function at(argv: (string | PathSpec)[], i: number): string | PathSpec {
  const v = argv[i]
  if (v === undefined) throw new CondError('argument expected')
  return v
}

async function evalTwo(ctx: CondContext, argv: (string | PathSpec)[]): Promise<boolean> {
  const first = scopePath(at(argv, 0))
  if (first === '!') return !evalOne(at(argv, 1))
  if (UNARY_OPS.has(first)) return applyUnary(ctx, first, at(argv, 1))
  throw new CondError(`${ctx.name}: ${first}: unary operator expected`)
}

async function evalThree(ctx: CondContext, argv: (string | PathSpec)[]): Promise<boolean> {
  const mid = scopePath(at(argv, 1))
  if (mid === '-a') {
    return evalOne(at(argv, 0)) && evalOne(at(argv, 2))
  }
  if (mid === '-o') {
    return evalOne(at(argv, 0)) || evalOne(at(argv, 2))
  }
  if (BINARY_OPS.has(mid)) {
    return applyBinary(ctx, at(argv, 0), mid, at(argv, 2))
  }
  const first = scopePath(at(argv, 0))
  if (first === '!') return !(await evalTwo(ctx, argv.slice(1)))
  if (first === '(' && scopePath(at(argv, 2)) === ')') {
    return evalOne(at(argv, 1))
  }
  throw new CondError(`${ctx.name}: ${mid}: binary operator expected`)
}

async function evalFour(ctx: CondContext, argv: (string | PathSpec)[]): Promise<boolean> {
  const first = scopePath(at(argv, 0))
  if (first === '!') return !(await evalThree(ctx, argv.slice(1)))
  if (first === '(' && scopePath(at(argv, 3)) === ')') {
    return evalTwo(ctx, argv.slice(1, 3))
  }
  return new ExprParser(ctx, argv).run()
}

/**
 * Recursive-descent `test` expression parser (>4 args, GNU expr grammar:
 * or -> and -> term with `!` and parentheses).
 */
class ExprParser {
  private pos = 0

  constructor(
    private readonly ctx: CondContext,
    private readonly argv: (string | PathSpec)[],
  ) {}

  private peek(offset = 0): string | null {
    const arg = this.argv[this.pos + offset]
    if (arg === undefined) return null
    return scopePath(arg)
  }

  async run(): Promise<boolean> {
    const result = await this.orExpr()
    if (this.pos !== this.argv.length) {
      throw new CondError(`${this.ctx.name}: too many arguments`)
    }
    return result
  }

  private async orExpr(): Promise<boolean> {
    let result = await this.andExpr()
    while (this.peek() === '-o') {
      this.pos += 1
      const right = await this.andExpr()
      result = result || right
    }
    return result
  }

  private async andExpr(): Promise<boolean> {
    let result = await this.term()
    while (this.peek() === '-a') {
      this.pos += 1
      const right = await this.term()
      result = result && right
    }
    return result
  }

  private async term(): Promise<boolean> {
    const tok = this.peek()
    if (tok === null) throw new CondError(`${this.ctx.name}: argument expected`)
    if (tok === '!') {
      this.pos += 1
      return !(await this.term())
    }
    if (tok === '(') {
      this.pos += 1
      const result = await this.orExpr()
      if (this.peek() !== ')') throw new CondError(`${this.ctx.name}: \`)' expected`)
      this.pos += 1
      return result
    }
    const nxt = this.peek(1)
    if (nxt !== null && BINARY_OPS.has(nxt) && this.peek(2) !== null) {
      const left = at(this.argv, this.pos)
      const right = at(this.argv, this.pos + 2)
      this.pos += 3
      return applyBinary(this.ctx, left, nxt, right)
    }
    if (UNARY_OPS.has(tok) && nxt !== null && nxt !== '-a' && nxt !== '-o') {
      const operand = at(this.argv, this.pos + 1)
      this.pos += 2
      return applyUnary(this.ctx, tok, operand)
    }
    this.pos += 1
    return tok !== ''
  }
}

/** Evaluate a flat test/[ argument list with bash's arity rules. */
async function evalFlat(ctx: CondContext, argv: (string | PathSpec)[]): Promise<boolean> {
  const n = argv.length
  if (n === 0) return false
  if (n === 1) return evalOne(at(argv, 0))
  if (n === 2) return evalTwo(ctx, argv)
  if (n === 3) return evalThree(ctx, argv)
  if (n === 4) return evalFour(ctx, argv)
  return new ExprParser(ctx, argv).run()
}

/** Evaluate a structured [[ ]] expression tree. */
async function evalCond(ctx: CondContext, node: CondNode): Promise<boolean> {
  if (node.kind === 'and') {
    return (await evalCond(ctx, node.left)) && (await evalCond(ctx, node.right))
  }
  if (node.kind === 'or') {
    return (await evalCond(ctx, node.left)) || (await evalCond(ctx, node.right))
  }
  if (node.kind === 'not') return !(await evalCond(ctx, node.inner))
  if (node.kind === 'unary') {
    if (!UNARY_OPS.has(node.op)) {
      throw new CondError('mirage: conditional unary operator expected')
    }
    return applyUnary(ctx, node.op, node.operand)
  }
  if (node.kind === 'binary') return evalCondBinary(ctx, node)
  return node.value !== ''
}

function evalCondBinary(ctx: CondContext, node: Extract<CondNode, { kind: 'binary' }>): boolean {
  if (node.op === '=' || node.op === '==') {
    if (node.rightLiteral) return node.left === node.right
    return fnmatch(node.left, node.right)
  }
  if (node.op === '!=') {
    if (node.rightLiteral) return node.left !== node.right
    return !fnmatch(node.left, node.right)
  }
  if (node.op === '=~') {
    const pattern = node.rightLiteral
      ? node.right.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
      : node.right
    let match: RegExpExecArray | null
    try {
      match = new RegExp(pattern).exec(node.left)
    } catch {
      throw new CondError('mirage: syntax error in conditional expression')
    }
    if (match === null) return false
    ctx.session.arrays.BASH_REMATCH = [
      match[0],
      ...match.slice(1).map((g: string | undefined) => g ?? ''),
    ]
    return true
  }
  if (node.op === '<') return node.left < node.right
  if (node.op === '>') return node.left > node.right
  if (NUMERIC_BINARY.has(node.op)) {
    // [[ evaluates numeric operands as arithmetic: variables resolve,
    // expressions compute, bare unset words are 0.
    let li: bigint
    let ri: bigint
    try {
      li = evaluateArith(node.left, ctx.session.env).value
      ri = evaluateArith(node.right, ctx.session.env).value
    } catch (exc) {
      if (!(exc instanceof ArithError)) throw exc
      throw new CondError('mirage: syntax error in conditional expression')
    }
    return compareInts(node.op, li, ri)
  }
  if (FILE_PAIR_BINARY.has(node.op)) {
    throw new CondError(`${ctx.name}: ${node.op}: unsupported operator`)
  }
  throw new CondError('mirage: conditional binary operator expected')
}

/**
 * Evaluate test/[ (flat argv) or [[ (condition tree). `name` is the
 * invocation spelling used in diagnostics: "test", "[", or "[[".
 */
export async function handleTest(
  dispatch: DispatchFn,
  namespace: Namespace,
  args: (string | PathSpec)[] | CondNode,
  session: Session,
  name = 'test',
): Promise<Result> {
  const ctx: CondContext = { dispatch, namespace, session, name }
  let result: boolean
  try {
    if (Array.isArray(args)) {
      result = await evalFlat(ctx, args)
    } else {
      result = await evalCond(ctx, args)
    }
  } catch (exc) {
    if (!(exc instanceof CondError)) throw exc
    const stderr = new TextEncoder().encode(exc.message + '\n')
    if (name === '[[') {
      // A bad [[ ]] operator is a bash PARSE error: the whole input
      // line dies, not just this command.
      throw new ExitSignal(2, stderr, null, 2)
    }
    return [
      null,
      new IOResult({ exitCode: 2, stderr }),
      new ExecutionNode({ command: 'test', exitCode: 2, stderr }),
    ]
  }
  const code = result ? 0 : 1
  return [
    null,
    new IOResult({ exitCode: code }),
    new ExecutionNode({ command: 'test', exitCode: code }),
  ]
}
